src/backend/distributed/commands/view.c

```diff
@@ -0,0 +1,434 @@
+static void AppendOptionsToCreateViewCommand(StringInfo createViewCommand, Oid viewOid);
+
+/*
+ * PreprocessViewStmt is called during the planning phase for CREATE OR REPLACE VIEW
+ * before it is created on the local node internally.
+ */
+List *
+PreprocessViewStmt(Node *node, const char *queryString,
+				   ProcessUtilityContext processUtilityContext)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	/* check creation against multi-statement transaction policy */
+	if (!ShouldPropagateCreateInCoordinatedTransction())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	return NIL;
+}
+
+
+/*
+ * PostprocessViewStmt actually creates the commmands we need to run on workers to
+ * propagate views.
+ *
+ * If view depends on any undistributable object, Citus can not distribute it. In order to
+ * not to prevent users from creating local views on the coordinator WARNING message will
+ * be sent to the customer about the case instead of erroring out. If no worker nodes exist
+ * at all, view will be created locally without any WARNING message.
+ *
+ * Besides creating the plan we also make sure all (new) dependencies of the view are
+ * created on all nodes.
+ */
+List *
+PostprocessViewStmt(Node *node, const char *queryString)
+{
+	ViewStmt *stmt = castNode(ViewStmt, node);
+
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	/* check creation against multi-statement transaction policy */
+	if (!ShouldPropagateCreateInCoordinatedTransction())
+	{
+		return NIL;
+	}
+
+	ObjectAddress viewAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+
+	if (IsObjectAddressOwnedByExtension(&viewAddress, NULL))
+	{
+		return NIL;
+	}
+
+	/* If the view has any unsupported dependency, create it locally */
+	DeferredErrorMessage *errMsg = DeferErrorIfHasUnsupportedDependency(&viewAddress);
+
+	if (errMsg != NULL)
+	{
+		/*
+		 * Don't need to give any warning/error messages if there is no worker nodes in
+		 * the cluster as user's experience won't be affected on the single node even
+		 * if the view won't be distributed.
+		 */
+		if (!HasAnyNodes())
+		{
+			return NIL;
+		}
+
+		/*
+		 * Since Citus drops and recreates views while converting a table type, giving a
+		 * NOTICE message is enough if the process in table type conversion function call
+		 */
+		if (InTableTypeConversionFunctionCall)
+		{
+			RaiseDeferredError(errMsg, NOTICE);
+			return NIL;
+		}
+
+		/*
+		 * If the view is already distributed, we should provide an error to not have
+		 * different definition of view on coordinator and worker nodes. If the view
+		 * is not distributed yet, we can create it locally to not affect user's local
+		 * usage experience.
+		 */
+		if (IsObjectDistributed(&viewAddress))
+		{
+			RaiseDeferredError(errMsg, ERROR);
+		}
+		else
+		{
+			RaiseDeferredError(errMsg, WARNING);
+			return NIL;
+		}
+	}
+
+	EnsureDependenciesExistOnAllNodes(&viewAddress);
+
+	char *command = CreateViewDDLCommand(viewAddress.objectId);
+
```

Comment:
needs comment why we use this trick and not `NodeDDLTaskList`. Please use the following as a baseline, and at least fix the grammar issues / typos etc.
```
/*
We'd typically use NodeDDLTaskList() for generating node-level 
ddl commands, such as when creating a type etc. However, 
views are different in a sense that views do not depend 
on citus tables. Instead, they are `depending` on citus tables.  

When NodeDDLTaskList() used, it should be accompanied 
with sequential execution.  Here, we do something equivalent 
to NodeDDLTaskList(), but using metadataSyncCommand field. 
This hack allows us to the metadata connection (see `REQUIRE_METADATA_CONNECTION` flag). Meaning that,
view creation is treated as a metadata operation. 

We do this mostly for performance reasons, because we cannot
afford to switch to sequential execution, for instance when we
are altering or creating distributed tables -- which may require 
significant resources.

The downside of using this hack is that if a view is re-used in 
the same transaction that creates the view on the workers, 
we might get errors such as the below -- which we consider 
a decent trade-off/limitation currently:

BEGIN;
    CREATE VIEW dist_view ..
    CRETAE TABLE t2(id int, val dist_view);

    -- shard creation fails on one of the connections
    SELECT create_distributed_table('t2', 'id');
    ERROR: type "public.dist_view_2" does not exist

*/
```